### PR TITLE
feat: create temporary directories

### PR DESCRIPTION
**Pull request by david-christiansen**
This PR adds a primitive for creating temporary directories, akin to the existing functionality for creating temporary files.

### CODE REVIEW

File: src/Init/System/IO.lean [hunk 1]
```diff
@@ -462,6 +462,17 @@ Note that it is the caller's job to remove the file after use.
 -/
 @[extern "lean_io_create_tempfile"] opaque createTempFile : IO (Handle × FilePath)
 
+/--
+Creates a temporary directory in the most secure manner possible. There are no race conditions in the
+file’s creation. The file is readable and writable only by the creating user ID. Additionally
```
Comment (hargoniX) [inline src/Init/System/IO.lean:467]: While directories are file's on UNIX we should probably refer to directory here? Also if the directory was not executable by anyone that would be rather bad. After all you need executable permissions to work with the directory.

Comment (david-christiansen) [reply src/Init/System/IO.lean]: Thank you - I believe the new version fixes this.

File: src/Init/System/IO.lean [hunk 2]
```diff
@@ -693,14 +694,38 @@ partial def FS.createDirAll (p : FilePath) : IO Unit := do
...
+def withTempFile [Monad m] [MonadFinally m] [MonadLiftT IO m] (f : Handle → FilePath → m α) :
+    m α := do
+  let (handle, path) ← createTempFile
+  try
+    f handle path
+  finally
+    removeFile path
+
+/--
+Like `createTempDir`, but also takes care of removing the directory after usage.
```
Comment (fgdorais) [inline src/Init/System/IO.lean:717]: For abundance of clarity, the docs should mention that everything in the temp directory will be deleted, including files added by the subprocess.

Comment (david-christiansen) [reply src/Init/System/IO.lean:717]: Done!

**Pull request merged**